Guide to Deploying a Contract from Installed Wasm Bytecode Using Code (#737)
stellar/stellar-docs: docs/build/guides/conventions/deploy-contracts-wasm.mdx

Reviewer: This is one of the changes that I'd like to see happen

##### **Function breakdown**

- `env: Env`: The Env object represents the current contract execution environment. It provides methods to interact with the blockchain and other contracts, as well as perform various operations.
- `deployer: Address`: The Address of the entity that is requesting the contract deployment. This address will be used to authorize the deployment.
- `wasm_hash: BytesN<32>`: The hash of the Wasm bytecode of the contract to be deployed. This hash is used to uniquely identify the contract's code.
- `salt: BytesN<32>`: A unique value used to derive the address of the deployed contract. The same combination of deployer address and salt will always produce the same deployed contract address, ensuring deterministic contract addresses.
- `init_fn: Symbol`: The name of the initialization function to be called on the newly deployed contract.
- `init_args: Vec<Val>`: A vector of arguments to be passed to the initialization function of the deployed contract.

> ##### **Function steps**

Reviewer: This is one of the changes that I'd like to see happen

```

- When you deploy a smart contract to a network, you need to specify an identity that will be used to sign the transactions. Change the `0xkodesage` to your own [identity].

[identity]: ../../smart-contracts/getting-started/setup.mdx#configure-an-identity

The command prints out the hash as hex. It will look something like `6bc6d975ef99c057231481c40f69f4c2a8a89ac56e8826ef0378f8e5c6abc4be`.

We also need to deploy the `Deployer` contract:

```sh
stellar contract deploy \
  --wasm deployer/target/wasm32-unknown-unknown/release/soroban_deployer_contract.wasm  --source 0xkodesage  \
  --network testnet
```

This will return the deployer address. For example: `CDKYZMA3OXR54YAHOQ5D4EWDFDT3ZNKKRYKQT7MGPWH22ZVD2DX2BJID`.

Then the deployer contract may be invoked with the Wasm hash value above.

```sh
stellar contract invoke \
  --id CDKYZMA3OXR54YAHOQ5D4EWDFDT3ZNKKRYKQT7MGPWH22ZVD2DX2BJID \
  --source 0xkodesage \
  --network testnet \
  -- \
  deploy \
  --deployer CDKYZMA3OXR54YAHOQ5D4EWDFDT3ZNKKRYKQT7MGPWH22ZVD2DX2BJID \
  --salt 123 \
  --wasm_hash 6bc6d975ef99c057231481c40f69f4c2a8a89ac56e8826ef0378f8e5c6abc4be \
  --init_fn init \
  --init_args '[{"u32":8}]'
```

> - replace `0xkodesage` with your own identity

Reviewer: This is one of the changes that I'd like to see happen

The deployer contract invocation will return the Contract address (For example: `CCTVFX6BFTQHTGAHA5TY4YZQJRUKRE2RRNUTGVBNKE3PJF5C7CI53APY`) of the newly deployed test contract.

Invoke the deployed test contract using the address returned from the previous command.

```sh
stellar contract invoke \
  --id CCTVFX6BFTQHTGAHA5TY4YZQJRUKRE2RRNUTGVBNKE3PJF5C7CI53APY \

>   --source 0xkodesage \

Reviewer: This is one of the changes that I'd like to see happen

    ///
    /// This has to be authorized by `deployer` (unless the `Deployer` instance
    /// itself is used as deployer). This way the whole operation is atomic
    /// and it's not possible to frontrun the contract initialization.
    ///
    /// Returns the contract address and result of the init function.
    pub fn deploy(
        env: Env,
        deployer: Address,
        wasm_hash: BytesN<32>,
        salt: BytesN<32>,
        init_fn: Symbol,
        init_args: Vec<Val>,
    ) -> (Address, Val) {
        // Skip authorization if deployer is the current contract.
        if deployer != env.current_contract_address() {
            deployer.require_auth();
        }

        // Deploy the contract using the uploaded Wasm with given hash.
        let deployed_address = env
            .deployer()
            .with_address(deployer, salt)
            .deploy(wasm_hash);

        // Invoke the init function with the given arguments.
        let res: Val = env.invoke_contract(&deployed_address, &init_fn, init_args);
        // Return the contract ID of the deployed contract and the result of
        // invoking the init result.
        (deployed_address, res)
    }
}

```

> #### How it works

Reviewer: This is one of the changes that I'd like to see happen

The deployer contract provides a mechanism to deploy other contracts in a secure and deterministic manner. It ensures that the deployment is authorized by the specified deployer address and supports atomic initialization of the newly deployed contract. This guarantees that the contract is properly initialized immediately after deployment, preventing any potential issues with uninitialized contracts.

> ##### **Function breakdown**

Reviewer: This is one of the changes that I'd like to see happen

```rust
(deployed_address, res)
```

- Returns a tuple containing the address of the newly deployed contract and the result of the initialization function.

### Build the contracts

To build the contract into a `.wasm` file, use the `stellar contract build` command. This command builds both the deployer contract and the test contract.

```sh
stellar contract build
```

Both `.wasm` files should be found in both contract `target` directories after building both contracts:

```
target/wasm32-unknown-unknown/release/soroban_deployer_contract.wasm
```

```
target/wasm32-unknown-unknown/release/soroban_deployer_test_contract.wasm
```

## Run the contract

Before deploying the test contract with the deployer, install the test contract Wasm using the `install` command. The `install` command will print out the hash derived from the Wasm file which should be used by the deployer.

```sh
stellar contract install --source-account 0xkodesage  --wasm  contract/target/wasm32-unknown-unknown/release/soroban_deployer_test_contract.wasm --rpc-url https://soroban-testnet.stellar.org:443 --network-passphrase "Test SDF Network ; September 2015"   --network testnet
```

> - When you deploy a smart contract to a network, you need to specify an identity that will be used to sign the transactions. Change the `0xkodesage` to your own [identity].

Reviewer: This is one of the changes that I'd like to see happen

[identity]: ../../smart-contracts/getting-started/setup.mdx#configure-an-identity

The command prints out the hash as hex. It will look something like `6bc6d975ef99c057231481c40f69f4c2a8a89ac56e8826ef0378f8e5c6abc4be`.

We also need to deploy the `Deployer` contract:

```sh
stellar contract deploy \
  --wasm deployer/target/wasm32-unknown-unknown/release/soroban_deployer_contract.wasm  --source 0xkodesage  \
  --network testnet
```

This will return the deployer address: `CDKYZMA3OXR54YAHOQ5D4EWDFDT3ZNKKRYKQT7MGPWH22ZVD2DX2BJID`.

Then the deployer contract may be invoked with the Wasm hash value above.

```sh
stellar contract invoke \
  --id CDKYZMA3OXR54YAHOQ5D4EWDFDT3ZNKKRYKQT7MGPWH22ZVD2DX2BJID \
  --source 0xkodesage \
  --network testnet \
  -- \
  deploy \
  --deployer CDKYZMA3OXR54YAHOQ5D4EWDFDT3ZNKKRYKQT7MGPWH22ZVD2DX2BJID \
  --salt 123 \
  --wasm_hash 6bc6d975ef99c057231481c40f69f4c2a8a89ac56e8826ef0378f8e5c6abc4be \
  --init_fn init \
  --init_args '[{"u32":8}]'
```

- replace `0xkodesage` with your own identity

it returns a Contract identifier `CCTVFX6BFTQHTGAHA5TY4YZQJRUKRE2RRNUTGVBNKE3PJF5C7CI53APY` of the deployed test contract.

> Invoke the deployed test contract using the identifier returned from the previous command.

Reviewer: ```suggestion
Invoke the deployed test contract using the address returned from the previous command.
```

```sh
stellar contract invoke \
  --id CCTVFX6BFTQHTGAHA5TY4YZQJRUKRE2RRNUTGVBNKE3PJF5C7CI53APY \
  --source 0xkodesage \
  --network testnet \

>     -- \
>     value

Reviewer: ```suggestion
  -- \
  value
```

```

> The following output should occur using the code above.

Reviewer: ```suggestion
You should receive something like the following output:
```

    ///
    /// This has to be authorized by `deployer` (unless the `Deployer` instance
    /// itself is used as deployer). This way the whole operation is atomic
    /// and it's not possible to frontrun the contract initialization.
    ///
    /// Returns the contract address and result of the init function.
    pub fn deploy(
        env: Env,
        deployer: Address,
        wasm_hash: BytesN<32>,
        salt: BytesN<32>,
        init_fn: Symbol,
        init_args: Vec<Val>,
    ) -> (Address, Val) {
        // Skip authorization if deployer is the current contract.
        if deployer != env.current_contract_address() {
            deployer.require_auth();
        }

        // Deploy the contract using the uploaded Wasm with given hash.
        let deployed_address = env
            .deployer()
            .with_address(deployer, salt)
            .deploy(wasm_hash);

        // Invoke the init function with the given arguments.
        let res: Val = env.invoke_contract(&deployed_address, &init_fn, init_args);
        // Return the contract ID of the deployed contract and the result of
        // invoking the init result.
        (deployed_address, res)
    }
}

> 

Reviewer: ```suggestion
```

Let's remove this empty line
